core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerPool.scala

```diff
@@ -303,26 +326,94 @@ class ContainerPool(childFactory: ActorRefFactory => ActorRef,
   }
 
   /** Install prewarm containers up to the configured requirements for each kind/memory combination. */
-  def backfillPrewarms(init: Boolean) = {
+  def adjustPrewarmedContainer(init: Boolean, scheduled: Boolean): Unit = {
     prewarmConfig.foreach { config =>
+      // Delete unused prewarmed container until minCount is reached
       val kind = config.exec.kind
       val memory = config.memoryLimit
-      val currentCount = prewarmedPool.count {
-        case (_, PreWarmedData(_, `kind`, `memory`, _)) => true //done starting
-        case _                                          => false //started but not finished starting
+
+      val runningCount = prewarmedPool.count {
+        // done starting, and not expired
+        case (_, p @ PreWarmedData(_, `kind`, `memory`, _, _)) if !p.isExpired() => true
+        // started but not finished starting (or expired)
+        case _ => false
       }
       val startingCount = prewarmStartingPool.count(p => p._2._1 == kind && p._2._2 == memory)
-      val containerCount = currentCount + startingCount
-      if (containerCount < config.count) {
+      val currentCount = runningCount + startingCount
+
+      logging.info(
+        this,
+        s"[kind: ${kind} memory: ${memory.toString}] currentCount: ${currentCount} prewarmed container")
+
+      // determine how many are needed
+      val desiredCount: Int =
+        if (init) config.initialCount
+        else {
+          if (scheduled) {
+            // scheduled/reactive config backfill
+            config.reactive
+              .map(c => getReactiveCold(c, kind, memory).getOrElse(c.minCount)) //reactive -> desired is either cold start driven, or minCount
+              .getOrElse(config.initialCount) //not reactive -> desired is always initial count
+          } else {
+            // normal backfill after removal - make sure at least minCount or initialCount is started
+            config.reactive.map(_.minCount).getOrElse(config.initialCount)
+          }
+        }
+
+      logging.info(
+        this,
+        s"[kind: ${kind} memory: ${memory.toString}] needs ${desiredCount} desired prewarmed container")
+
+      // remove expired
+      config.reactive.foreach { _ =>
+        val expiredPrewarmedContainer = prewarmedPool
+          .filter { warmInfo =>
+            warmInfo match {
+              case (_, p @ PreWarmedData(_, `kind`, `memory`, _, _)) if p.isExpired() => true
+              case _                                                                  => false
+            }
+          }
+        // emit expired container counter metric with memory + kind
+        val removedCount = expiredPrewarmedContainer.size
+        MetricEmitter.emitHistogramMetric(LoggingMarkers.CONTAINER_POOL_EXPIRED(memory.toString, kind), removedCount)
```

Comment:
I would name this metric "prewarmExpired"

Reply:
Also, these metrics should be counters not histograms, I think.

Reply:
updated accordingly